src/proxy/server/downstream.rs

```diff
@@ -0,0 +1,203 @@
+use prometheus::HistogramTimer;
+use tokio::sync::{mpsc, watch};
+use tokio::time::Duration;
+use tracing::Instrument;
+
+use crate::{
+    cluster::cluster_manager::SharedClusterManager,
+    endpoint::{Endpoint, EndpointAddress},
+    filters::{manager::SharedFilterManager, Filter, ReadContext},
+    proxy::sessions::{
+        metrics::Metrics as SessionMetrics, session_manager::SessionManager, Session, SessionArgs,
+        SessionKey, UpstreamPacket,
+    },
+};
+
+use super::metrics::Metrics as ProxyMetrics;
+
+/// Packet received from local port
+#[derive(Debug)]
+pub(crate) struct DownstreamPacket {
+    pub source: EndpointAddress,
+    pub contents: Vec<u8>,
+    pub timer: HistogramTimer,
+}
+
+/// Represents the required arguments to run a worker task that
+/// processes packets received downstream.
+pub(crate) struct DownstreamReceiveWorker {
+    /// ID of the worker.
+    pub worker_id: usize,
+    /// Channel from which the worker picks up the downstream packets.
+    pub packet_rx: mpsc::Receiver<DownstreamPacket>,
+    /// Configuration required to process a received downstream packet.
+    pub processor: DownstreamReceiveProcessor,
+    /// The worker task exits when a value is received from this shutdown channel.
+    pub shutdown_rx: watch::Receiver<()>,
+}
+
+impl DownstreamReceiveWorker {
+    // For each worker config provided, spawn a background task that sits in a
+    // loop, receiving packets from a queue and processing them through
+    // the filter chain.
+    #[tracing::instrument(skip_all, fields(worker_id))]
+    pub async fn run(self) {
+        let Self {
+            worker_id,
+            mut packet_rx,
+            mut shutdown_rx,
+            processor,
+        } = self;
+
+        loop {
+            tokio::select! {
+                packet = packet_rx.recv().instrument(tracing::trace_span!("Received downstream packet")) => {
```

Comment:
I'm learning lots of neat things about the tracing library. That's super nice that tracing allows you to attach the instrumentation to the Future.